proposals/testing/NNNN-polling-confirmations.md

```diff
@@ -0,0 +1,512 @@
+) async throws -> R
+```
+
+### New `PollingStopCondition` enum
+
+A new enum type, `PollingStopCondition` will be defined, specifying when to stop
+polling before the duration has elapsed. Additionally, if the early stop
+condition isn't fulfilled before the duration elapses, then this also defines
+how the confirmation should be handled.
+
+```swift
+/// A type defining when to stop polling early.
+/// This also determines what happens if the duration elapses during polling.
+public enum PollingStopCondition: Sendable {
+  /// Evaluates the expression until the first time it returns true.
+  /// If it does not pass once by the time the timeout is reached, then a
+  /// failure will be reported.
+  case firstPass
+
+  /// Evaluates the expression until the first time it returns false.
+  /// If the expression returns false, then a failure will be reported.
+  /// If the expression only returns true before the timeout is reached, then
+  /// no failure will be reported.
+  /// If the expression does not finish evaluating before the timeout is
+  /// reached, then a failure will be reported.
+  case stopsPassing
+}
+```
+
+### New Error Type
+
+A new error type, `PollingFailedError` to be thrown when the polling
+confirmation doesn't pass:
+
+```swift
+/// A type describing an error thrown when polling fails.
+public struct PollingFailedError: Error, Sendable, CustomIssueRepresentable {}
+```
+
+### New `Issue.Kind` case
+
+A new issue kind will be added to report specifically when a test fails due to
+a failed polling confirmation.
+
+```swift
+public struct Issue {
+  public enum Kind {
+    /// An issue due to a polling confirmation having failed.
+    ///
+    /// This issue can occur when calling ``confirmation(_:until:within:pollingEvery:isolation:sourceLocation:_:)-455gr``
+    /// or
+    /// ``confirmation(_:until:within:pollingEvery:isolation:sourceLocation:_:)-5tnlk``
+    /// whenever the polling fails, as described in ``PollingStopCondition``.
+    case pollingConfirmationFailed
+  }
+}
+```
+
+### New Traits
+
+Two new traits will be added to change the default values for the
+`duration` and `interval` arguments. Test authors will often want to poll for
+the `firstPass` stop condition for longer than they poll for the
+`stopsPassing` stop condition, which is why there are separate traits for
+configuring defaults for these functions.
+
+```swift
+/// A trait to provide a default polling configuration to all usages of
+/// ``confirmation(_:until:within:pollingEvery:isolation:sourceLocation:_:)-455gr``
+/// and
+/// ``confirmation(_:until:within:pollingEvery:isolation:sourceLocation:_:)-5tnlk``
+/// within a test or suite for the ``PollingStopCondition.firstPass``
+/// stop condition.
+///
+/// To add this trait to a test, use the
+/// ``Trait/pollingUntilFirstPassDefaults`` function.
+@available(_clockAPI, *)
+public struct PollingUntilFirstPassConfigurationTrait: TestTrait, SuiteTrait {
```

Comment:
Assuming you do want to include the ability to influence this feature's behavior via a trait, I would suggest consolidating these two trait types into one if possible. For other traits we've tried to have the types be broader and use the `static` functions which initialize those types be the point where we differentiate their behavior. For example, the `.enabled(if:)` and `.disabled(if:)` functions both return a type of the same type: `ConditionTrait`.

But to accomplish this consolidation you'll likely need to store the `PollingStopCondition` in the trait instance. I think these changes would make for a simpler trait story for this feature: a single trait with a simpler name like `PollingConfigurationTrait`, with a property specifying its stopping condition.